arkane/ess/factory.py

```diff
@@ -0,0 +1,229 @@
+###############################################################################
+
+"""
+A module for generating ESS adapters
+"""
+
+import logging
+import os
+from typing import Type
+
+from .adapter import ESSAdapter
+
+from rmgpy.exceptions import InputError
+from rmgpy.qm.qmdata import QMData
+from rmgpy.qm.symmetry import PointGroupCalculator
+
+_registered_ess_adapters = {}
+
+
+def register_ess_adapter(ess: str,
+                         ess_class: Type[ESSAdapter],
+                         ) -> None:
+    """
+    A register for the ESS adapters.
+
+    Args:
+        ess: A string representation for an ESS adapter
+        ess_class: The ESS adapter class
+    Raises:
+        TypeError: If ess_class is not an ESSAdapter instance.
+    """
+    if not issubclass(ess_class, ESSAdapter):
+        raise TypeError(f'{ess_class} is not an ESSAdapter')
+    _registered_ess_adapters[ess] = ess_class
+
+
+def ess_factory(fullpath: str) -> Type[ESSAdapter]:
+    """
+    A factory generating the ESS adapter corresponding to ``ess_adapter``.
+    Given a path to the log file of a QM software, determine whether it is
+    Gaussian, Molpro, QChem, Orca, or TeraChem
+
+    Args:
+        fullpath (str): The disk location of the output file of interest.
+
+    Returns:
+        Type[ESSAdapter]: The requested ESSAdapter child, initialized with the respective arguments.
+    """
+
+    ess_name = None
+    with open(fullpath, 'r') as f:
+        if os.path.splitext(fullpath)[1] in ['.xyz', '.dat', '.geometry']:
+            ess_name = 'terachem'
+        line = f.readline()
+        while ess_name is None and line != '':
+            if 'gaussian' in line.lower():
+                ess_name = 'gaussian'
+                break
+            elif 'molpro' in line.lower():
+                ess_name = 'molpro'
+                break
+            elif 'O   R   C   A' in line or 'orca' in line.lower():
+                f.close()
+                ess_name = 'orca'
+                break
+            elif 'qchem' in line.lower():
+                ess_name = 'qchem'
+                break
+            elif 'terachem' in line.lower():
+                ess_name = 'terachem'
+                break
+            line = f.readline()
+    if ess_name is None:
+        raise InputError(f'The file at {fullpath} could not be identified as a '
+                         f'Gaussian, Molpro, Orca, QChem, or TeraChem log file.')
+
+    return _registered_ess_adapters[ess_name](path=fullpath)
+
+
+class Log(object):
+    def __init__(self, fullpath: str):
+        """
+        This class is not a part of the Factory/Adapter design class. It was added here to keep Arkane
+        backward compatible with current input files that use ``Log`` calls. Added a .ess attribute that stores
+        the specific ESS Adapter, along with shortcut methods to call the respective Adapter method.
+        """
+        self.path = fullpath
+        if os.path.isfile(self.path):
```

Comment:
There are various if statements in `arkane/statmech.py` that update fullpath if it was just passed a filename rather than an actual path. Calling `ess_factory` on just a filename, rather than fullpath, will throw an error. However, since the previous Log.py file would just accept a filename and let statmech.py update the fullpath, I was going to follow that trend. Do you want me to add an `else` statement that tries to change the filename to a fullpath here in Log.py or let statmech.py continue to handle that?

Reply:
Upon testing, the line `directory = os.path.abspath(os.path.dirname(path))` in `load` in `statmech.py` correctly returns a directory such as `/Users/kevin/Documents/Green_Lab/RMG/RMG-Py/examples/arkane/species/Benzyl`. However, using that command in factory.py returns nothing for me. I'm currently unsure why it doesn't behave as expected when called in factory.py but I'd recommend letting statmech.py continue to handle it since it works properly

 ```
       self.path = fullpath

        if os.path.isfile(self.path):

            self.ess = ess_factory(self.path)

        else:

            # fullpath is likely just a filename that must be made into a fullpath 

            print(f'directory name is: {os.path.abspath(os.path.dirname(fullpath))}')

            new_path = os.path.join(os.path.abspath(os.path.dirname(fullpath)), fullpath)

            self.ess = ess_factory(new_path)
```

Reply:
I see. Ideally, it would be great to have the rel -> absolute path correction done here instead of in statmech, but it's not a high priority. So up to you (OK to leave as is w/o `else`)